File: models/src/main/java/io/xj/hub/music/NoteRange.java
// Copyright (c) XJ Music Inc. (https://xj.io) All Rights Reserved.

package io.xj.hub.music;

import jakarta.annotation.Nullable;
import java.util.Collection;
import java.util.Objects;
import java.util.Optional;

/**
 * Represent a note range
 */
public class NoteRange {
  static final String UNKNOWN = "Unknown";
  static final int MAX_SEEK_OCTAVES = 3;

  @Nullable
  Note low;

  @Nullable
  Note high;

  @Nullable
  Integer span;

  NoteRange() {
    low = null;
    high = null;
    span = 0;
  }

  NoteRange(@Nullable Note low, @Nullable Note high) {
    this.low = low;
    this.high = high;
    span = Objects.nonNull(low) && Objects.nonNull(high) ? low.delta(high) : null;
  }

  NoteRange(@Nullable String low, @Nullable String high) {
    this.low = Objects.nonNull(low) ? Note.of(low) : null;
    this.high = Objects.nonNull(high) ? Note.of(high) : null;
    span = Objects.nonNull(low) && Objects.nonNull(high) ? this.low.delta(this.high) : null;
  }

  public static NoteRange from(@Nullable Note low, @Nullable Note high) {
    return new NoteRange(low, high);
  }

  public static NoteRange from(@Nullable String low, @Nullable String high) {
    return new NoteRange(low, high);
  }

  public static NoteRange copyOf(NoteRange range) {
    return new NoteRange(range.low, range.high);
  }

  public static NoteRange ofNotes(Collection<Note> notes) {
    return new NoteRange(notes.parallelStream().min(Note::compareTo).orElse(null), notes.parallelStream().max(Note::compareTo).orElse(null));
  }

  public static NoteRange ofStrings(Collection<String> notes) {
    return new NoteRange(notes.parallelStream().map(Note::of).min(Note::compareTo).orElse(null), notes.parallelStream().map(Note::of).max(Note::compareTo).orElse(null));
  }

  public static NoteRange median(NoteRange r1, NoteRange r2) {
    return new NoteRange(Note.median(r1.getLow().orElse(null), r2.getLow().orElse(null)), Note.median(r1.getHigh().orElse(null), r2.getHigh().orElse(null)));
  }

  public static NoteRange empty() {
    return new NoteRange();
  }

  /**
   * Compute the median optimal range shift octaves
   *
   * @param sourceRange from
   * @param targetRange to
   * @return median optimal range shift octaves
   */
  public static Integer computeMedianOptimalRangeShiftOctaves(NoteRange sourceRange, NoteRange targetRange) throws MusicalException {
    if (sourceRange.getLow().isEmpty() || sourceRange.getHigh().isEmpty() || targetRange.getLow().isEmpty() || targetRange.getHigh().isEmpty())
      return 0;
    var shiftOctave = 0; // search for optimal value
    var baselineDelta = 100; // optimal is the lowest possible integer zero or above
    for (var o = 10; o >= -10; o--) {
      int dLow = targetRange.getLow().orElseThrow(() -> new MusicalException("Can't find low end of target range")).delta(sourceRange.getLow().orElseThrow(() -> new MusicalException("Can't find low end of source range")).shiftOctave(o));
      int dHigh = targetRange.getHigh().orElseThrow(() -> new MusicalException("Can't find high end of target range")).delta(sourceRange.getHigh().orElseThrow(() -> new MusicalException("Can't find high end of source range")).shiftOctave(o));
      if (0 <= dLow && 0 >= dHigh && Math.abs(o) < baselineDelta) {
        baselineDelta = Math.abs(o);
        shiftOctave = o;
      }
    }
    return shiftOctave;
  }

  public Optional<Note> getLow() {
    return Optional.ofNullable(low);
  }

  public Optional<Note> getHigh() {
    return Optional.ofNullable(high);
  }

  public Optional<Integer> getSpan() {
    return Optional.ofNullable(span);
  }

  public String toString(Accidental accidental) {
    if (Objects.nonNull(low) && Objects.nonNull(high))
      return String.format("%s-%s", low.toString(accidental), high.toString(accidental));
    if (Objects.nonNull(low)) return String.format("%s", low.toString(accidental));
    if (Objects.nonNull(high)) return String.format("%s", high.toString(accidental));
    return UNKNOWN;
  }

  public void expand(Collection<Note> notes) {
    for (var note : notes) expand(note);
  }

  public void expand(Note note) {
    if (Objects.isNull(low) || note.isLower(low)) low = note;
    if (Objects.isNull(high) || note.isHigher(high)) high = note;
  }

  public void expand(NoteRange range) {
    if (Objects.nonNull(range.low)) expand(range.low);
    if (Objects.nonNull(range.high)) expand(range.high);
  }

  public int getDeltaSemitones(NoteRange target) {
    var s = getMedianNote();
    var t = target.getMedianNote();
    if (s.isEmpty() || t.isEmpty()) return 0;
    return s.get().delta(t.get());
  }

  public Optional<Note> getMedianNote() {
    if (Objects.isNull(low) && Objects.isNull(high)) return Optional.empty();
    if (Objects.isNull(low)) return Optional.of(high);
    if (Objects.isNull(high)) return Optional.of(low);
    return Optional.of(low.shift(low.delta(high) / 2));
  }

  public NoteRange shifted(int inc) {
    return new NoteRange(Objects.nonNull(low) ? low.shift(inc) : null, Objects.nonNull(high) ? high.shift(inc) : null);
  }

  public boolean isEmpty() {
    return Objects.isNull(low) || Objects.isNull(high) || PitchClass.None.equals(low.getPitchClass()) || PitchClass.None.equals(high.getPitchClass());
  }

  public Optional<Note> getNoteNearestMedian(PitchClass root) {
    if (PitchClass.None.equals(root)) return Optional.empty();
    var median = getMedianNote();
    if (median.isEmpty()) return Optional.empty();
    if (Objects.equals(root, median.get().getPitchClass())) return median;
    var up = median.get().nextUp(root);
    var down = median.get().nextDown(root);
    return down.delta(median.get()) < median.get().delta(up) ? Optional.of(down) : Optional.of(up);
  }

  /**
   * Change the octave of a note such that it is within this range
   *
   * @param note source
   * @return note moved to available octave
   */
  public Note toAvailableOctave(Note note) {
    if (Objects.isNull(low) || Objects.isNull(high)) return note;

    int d = 0;
    Note x = note;

    while (!includes(x) && d < MAX_SEEK_OCTAVES) {
      if (low.isHigher(x)) {
        x = x.shiftOctave(1);
        d++;
      } else if (high.isLower(x)) {
        x = x.shiftOctave(-1);
        d++;
      }
    }

    return x;
  }

  /**
   * Whether the given note is within this range
   *
   * @param note to test
   * @return true if note is within this range
   */
  public boolean includes(Note note) {
    if (Objects.isNull(low) && Objects.isNull(high)) return false;
    if (Objects.isNull(low) && high.sameAs(note)) return true;
    if (Objects.isNull(high) && low.sameAs(note)) return true;
    if (Objects.isNull(low) || Objects.isNull(high)) return false;
    return !low.isHigher(note) && !high.isLower(note);
  }
}

File: engine/src/music/NoteRange.cpp
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.

#include <sstream>

#include "xjmusic/music/NoteRange.h"

using namespace XJ;


NoteRange::NoteRange() : low(std::nullopt), high(std::nullopt) {}


NoteRange::NoteRange(const std::optional<Note> low, const std::optional<Note> high) : low(low),
                                                                          high(high) {
}


NoteRange::NoteRange(const std::string &low, const std::string &high) : NoteRange(Note::ifTonal(low),
                                                                                  Note::ifTonal(high)) {}


NoteRange NoteRange::from(Note low, Note high) {
  return {
      low.isAtonal() ? std::nullopt : std::optional(low),
      high.isAtonal() ? std::nullopt : std::optional(high)
  };
}


NoteRange NoteRange::from(const std::string &low, const std::string &high) {
  return {low, high};
}


NoteRange NoteRange::copyOf(const NoteRange &range) {
  return {range.low, range.high};
}


NoteRange NoteRange::ofNotes(std::vector<Note> notes) {
  const std::set<Note> properNotes(notes.begin(), notes.end());
  return ofNotes(properNotes);
}


NoteRange NoteRange::ofNotes(const std::set<Note>& notes) {
  const auto minNote = std::min_element(notes.begin(), notes.end());
  const auto maxNote = std::max_element(notes.begin(), notes.end());
  std::optional<Note> low = minNote != notes.end() ? std::optional(*minNote) : std::nullopt;
  std::optional<Note> high = maxNote != notes.end() ? std::optional(*maxNote) : std::nullopt;
  return {low, high};
}


NoteRange NoteRange::ofStrings(const std::vector<std::string> &notes) {
  std::set<Note> properNotes;
  for (const auto &note: notes) {
    auto n = Note::ifValid(note);
    if (n.has_value()) properNotes.emplace(n.value());
  }
  return ofNotes(properNotes);
}


NoteRange NoteRange::median(const NoteRange &r1, const NoteRange &r2) {
  return {Note::median(r1.low, r2.low), Note::median(r1.high, r2.high)};
}


int NoteRange::computeMedianOptimalRangeShiftOctaves(const NoteRange *sourceRange, const NoteRange *targetRange) {
  if (!sourceRange->low.has_value() || !sourceRange->high.has_value() ||
      !targetRange->low.has_value() || !targetRange->high.has_value())
    return 0;

  int shiftOctave = 0;    // search for optimal value
  int baselineDelta = 100;// optimal is the lowest possible integer zero or above

  for (int o = 10; o >= -10; o--) {
    if (!targetRange->low.has_value())
      throw std::runtime_error("Can't find low end of target range");
    if (!sourceRange->low.has_value())
      throw std::runtime_error("Can't find low end of source range");
    const int dLow = targetRange->low->delta(sourceRange->low->shiftOctave(o));

    if (!targetRange->high.has_value())
      throw std::runtime_error("Can't find high end of target range");
    if (!sourceRange->high.has_value())
      throw std::runtime_error("Can't find high end of source range");
    const int dHigh = targetRange->high->delta(sourceRange->high->shiftOctave(o));

    if (0 <= dLow && 0 >= dHigh && std::abs(o) < baselineDelta) {
      baselineDelta = std::abs(o);
      shiftOctave = o;
    }
  }

  return shiftOctave;
}


std::string NoteRange::toString(const Accidental accidental) {
  std::stringstream ss;
  if (low.has_value() && high.has_value()) {
    ss << low->toString(accidental) << "-" << high->toString(accidental);
    return ss.str();
  }
  if (low.has_value()) {
    ss << low->toString(accidental);
    return ss.str();
  }
  if (high.has_value()) {
    ss << high->toString(accidental);
    return ss.str();
  }
  return UNKNOWN;
}


void NoteRange::expand(const std::vector<Note> &notes) {
  for (const auto note: notes) expand(note);
}


void NoteRange::expand(Note note) {
  if (!low.has_value() || note < low) low = note;
  if (!high.has_value() || note > high) high = note;
}


void NoteRange::expand(const NoteRange *range) {
  if (range->low.has_value()) expand(range->low.value());
  if (range->high.has_value()) expand(range->high.value());
}


int NoteRange::getDeltaSemitones(NoteRange target) {
  const auto s = getMedianNote();
  const auto t = target.getMedianNote();
  if (!s.has_value() || !t.has_value()) return 0;
  return s.value().delta(t.value());
}


std::optional<Note> NoteRange::getMedianNote() {
  if (!low.has_value() && !high.has_value()) return std::nullopt;
  if (!low.has_value()) return {high};
  if (!high.has_value()) return {low};
  return {low->shift(low->delta(high.value()) / 2)};
}


NoteRange NoteRange::shifted(const int inc) const {
  return {
      low.has_value() ? std::optional(low->shift(inc)) : std::nullopt,
      high.has_value() ? std::optional(high->shift(inc)) : std::nullopt};
}


bool NoteRange::empty() const {
  return !low.has_value() || !high.has_value() || Atonal == low->pitchClass ||
         Atonal == high->pitchClass;
}


std::optional<Note> NoteRange::getNoteNearestMedian(const PitchClass root) {
  if (Atonal == root) return std::nullopt;
  auto median = getMedianNote();
  if (!median.has_value()) return std::nullopt;
  if (root == median->pitchClass) return median;
  auto up = median->nextUp(root);
  auto down = median->nextDown(root);
  return down.delta(median.value()) < median->delta(up) ? std::optional(down) : std::optional(up);
}


Note NoteRange::toAvailableOctave(const Note note) const {
  if (!low.has_value() || !high.has_value()) return note;

  int d = 0;
  Note x = note;

  while (!includes(x) && d < MAX_SEEK_OCTAVES) {
    if (low > x) {
      x = x.shiftOctave(1);
      d++;
    } else if (high < x) {
      x = x.shiftOctave(-1);
      d++;
    }
  }

  return x;
}


bool NoteRange::includes(const Note note) const {
  if (!low.has_value() && !high.has_value()) return false;
  if (!low.has_value() && high.value() == note) return true;
  if (!high.has_value() && low.value() == note) return true;
  if (!low.has_value() || !high.has_value()) return false;
  return low.value() <= note && high.value() >= note;
}